### ISSUE TITLE
YACC does not work on Git 2.11 or higher

### ISSUE THREAD

**Issue by bturner** [sford/yet-another-commit-checker#139]
I'm the principal developer for Bitbucket Server. Git 2.11 makes significant changes to push handling on the server, which breaks `PreReceiveHook` and `PreReceiveRepositoryHook` implementations on Bitbucket Server. I made changes in Bitbucket Server 4.13 fix hooks without requiring any changes to the hooks themselves _unless the hooks use JGit_. It's not possible for me to transparently fix such hooks on the Bitbucket Server side. (This is all tracked on [BSERV-9388](https://jira.atlassian.com/browse/BSERV-9388))

Since YACC uses JGit to retrieve commit details to access the committer, this plugin still does not, and cannot, work with Git 2.11 or newer. The only way to fix the hook is to switch. You can use the `GitScmCommandBuilder` to build your own command for retrieving commit details.

From looking at how this plugin is implemented, I'd either use `git cat-file --batch` to access the details without needing to fork off a separate Git command for each, or I'd drop use of Bitbucket Server's `CommitService` and use `GitScmCommandBuilder` to build my own `git rev-list` command which included the committer details and parse my own output. Since `git rev-list` is what `CommitService.getCommitsBetween` uses under the covers (you really ought to switch to `streamCommitsBetween`, by the way; it'll make your hook significantly faster), switching to your own `git rev-list` command, while a little more effort to parse, wouldn't be any less performant.

**Comment by bturner** [comment]
Please let me know if you have questions. I'm happy to do what I can to help.

**Comment by sford** [comment]
Hi @bturner, thanks for reaching out! I just recently become aware of the issue (#137) and I am currently in the progress of moving YACC over to `GitScm` and `GitScmCommandBuilder` instead of using JGit. Those utilities are much better than what YACC is currently doing :-)

I will let you know if I have any questions. Thank you!

**Comment by bturner** [comment]
@sford Great! I appreciate your efforts; this is an add-on used by a lot of organizations.

A couple of thoughts, just while I'm here (and again, happy to answer any questions, either here or via e-mail if you want to reach out directly):

* You should never need to reference `GitScm` directly. You can `<component-import interface="com.atlassian.bitbucket.scm.git.command.GitCommandBuilderFactory"/>` and get the builder factory directly; you don't need to import `GitScm` to get the factory
* In your `pom.xml` you've got a dependency on `stash-scm-git-api` which is not `<scope>provided</scope>`. That will bundle our API into your plugin, which shouldn't happen. Are you sure you need that dependency? `bitbucket-git-api` (note that "scm" was dropped) is the 4.x version of the same thing
* Your plugin is referencing Bitbucket Server 4.0.0. I'd encourage you to upgrade this to 4.0.7. This won't change your supported versions (you'll still support all the way back to 4.0.0; patch releases of Bitbucket Server never include API or SPI changes), but you'll get a fix for an issue in the Bitbucket Server 4.0.0 POM which is bundling the `commons-lang` jar into your plugin. (See [BSERV-8228](https://jira.atlassian.com/browse/BSERV-8228))

**Comment by gsylvie** [comment]
This thread should be recommended reading for all Bitbucket add-on developers!   Very helpful!  (I maintain [Bit-Booster for Bitbucket Server](https://marketplace.atlassian.com/plugins/com.bit-booster.bb/server/overview)).

**Comment by gsylvie** [comment]
p.s.  @bturner , I actually want to move in the opposite direction and use JGit for more of my git operations.  If I wanted to use JGit in my hooks, where are the pushed objects stored now in Git 2.11?

Oh, and one more question.  Will this still be okay?

1. gitScm.getCommandBuilderFactory().builder().command("log").workingDirectory(gitDir)

As opposed to this invocation that provides a "Repository" object:

2. getCommandBuilderFactory().builder(repo).command("log")


Right now all my hooks use style #2, but my Rebase / Squash / Amend buttons use style #1, since they are maintaining their own clone in a different area on disk and managing ".git/objects/info/alternates" directly.   Obviously Rebase/ Squash / Amend are not going to be affected by changes to git's push behaviour, so really I'm just asking out of curiosity.

**Comment by bturner** [comment]
@gsylvie It is not possible to use JGit _at all_ in any `PreReceiveHook` or `PreReceiveRepositoryHook` implementations if the server is using Git 2.11 or newer. The new quarantining behavior that was added for pushes creates a random temporary directory where the newly-received objects are written, and exports `GIT_OBJECT_DIRECTORY`, `GIT_ALTERNATE_OBJECT_DIRECTORIES` and `GIT_QUARANTINE_PATH`. However, the `PreReceiveHook` and `PreReceiveRepositoryHook` SPIs in Bitbucket Server have no mechanism for providing environment variables to implementors. That means it's not possible for you to get those values and use them yourself to wire up JGit's objects. If you are only creating a `PostReceiveHook` or `AsyncPostReceiveRepositoryHook`, all of the received objects have been moved from the quarantine location into the real repository, so JGit should work normally in those hooks.

I'm investigating some potential changes to the hook SPIs in Bitbucket Server 5.0, which would allow me to expose environment variables (and potentially some other data) to hook implementations, but on 4.x it's not possible to get access to the information you need to build a working `PreReceiveHook` using anything other than Bitbucket Server's own APIs.

I'll note that JGit is...shall we say extremely liberal in its memory usage, especially for `RevWalk` operations (where it caches every `Commit` it reads in memory), and it has some other undesirable behaviors (like writing loose objects for objects it delta-decompresses, in some cases). It's possible some of these behaviors have changed in more recent versions, but in general I'd say you're better off using `git`, rather than JGit.

`builder(Repository)` is a convenient shorthand for `builder().workingDirectory(repositoryDir)`, so the two are equivalent in their behavior, but they're _not_ quite the same in their error handling. The `CommandErrorHandler` used by default, if you don't provide your own, uses the `Repository` to provide more exact error details, if one is available. So, to answer your question, you're welcome to use `builder().workingDirectory` instead of `builder(repository)`, but error messages may be slightly different.

**Comment by sford** [comment]
@bturner Thanks a lot for all the info and pointers. I got the initial change in to replace JGit with a call to `git rev-list` and parsing out the information I need. Overall, I think it is _much_ better. I still need to do some battle testing, and a bit more work to deal with tags. If you have time, I would appreciate a quick spot check on my [GitScmCommandBuilder usage](https://github.com/sford/yet-another-commit-checker/blob/master/src/main/java/com/isroot/stash/plugin/CommitsServiceImpl.java#L78-L91) to see if I am following best practices using the BitBucket API/SPI.

I couldn't quite get `GitCommandBuilderFactory` component import working... I was getting injection errors when BBS tried to load the add-on. `ScmService` worked for me though. I was going through a lot of iterations trying to get things working, so I might have been hitting another issue. The initial plugin structure for YACC was generated back in the Stash 2.8.2 days so part of me wonders if I am not following the latest and greatest BBS 4 recommended add-on structure (like with the pom.xml, spring injections, etc).

**Comment by gsylvie** [comment]
getCommandBuilderFactory().builder(repo).command("for-each-ref") with no arguments is one way to figure out which tags are lightweight and which are annotated:

```
36363359f8cba1fceb93c4b7084af235d543d745 commit	refs/heads/release/1.x
f22eb1e5d60ac36c5b793257678ed5cf6209fa5f commit	refs/tags/1.1
9640db61f9b93d00bf16fda145bd588977c31945 commit	refs/tags/E
47c79748ec6de88969b847e9f9510256b9886b60 tag	refs/tags/F
```

**Comment by bturner** [comment]
@sford I'm going to fork this repository and push a couple things to my fork. It might be a good way to share some ideas and insight into our API/SPI with you. One thing you'll need to decide, though, is whether you want to raise the minimum supported version for your plugin or not. The change you linked me to bumps your Bitbucket Server dependency to 4.13. At that point it's really not "safe" to say you support 4.0 (unless you're explicitly deploying to 4.0.x somewhere to verify it). If you want to support 4.x, all versions, you'll want to leave your dependency at 4.0.7.

@gsylvie Depending on the supported version range for your plugin, the absolute best thing you can do for deciding between lightweight and annotated tags is to use Bitbucket Server's API, _not_ `git`. `RefService.resolveRefs`, introduced in Bitbucket Server 4.6, is heavily optimized to be able to resolve refs using the fewest `git` processes possible (newer versions of Bitbucket Server have progressively optimized it; it's good in 4.6 but it's even better in 4.13).

If you're supporting versions of Bitbucket Server prior to 4.6, I'd still suggest using `RefService.resolveRef` (singular), which has existed since before 4.0. Note that Bitbucket Server 5.0 will remove `resolveRef`, though, in favor of the bulk interface.

As for using `git for-each-ref`, `refs/heads/**` may not have consistent behavior, but `refs/heads/` should; we have unit tests which explicitly verify this behavior from Git 1.8 all the way up to Git 2.11:

```
bturner@elysoun /tmp/foo.git (BARE:master)$ /opt/git/1.8.0/bin/git for-each-ref refs/heads/\*\*
25c6faff580f6e19490aa0eb320f0956faf1670c commit	refs/heads/master
bturner@elysoun /tmp/foo.git (BARE:master)$ /opt/git/1.8.0/bin/git for-each-ref refs/heads/
25c6faff580f6e19490aa0eb320f0956faf1670c commit	refs/heads/master
25c6faff580f6e19490aa0eb320f0956faf1670c commit	refs/heads/release/1.1.x/bob
25c6faff580f6e19490aa0eb320f0956faf1670c commit	refs/heads/release/1.2.x
```

The escaping in that first command is to stop Bash from expanding the arguments, which makes that `git for-each-ref` _appear_ to return the right data when in practice it wouldn't (Bitbucket Server _never_ uses a shell to execute `git` on any platform).

**Comment by sford** [comment]
I just released YACC 1.13 with jGit completely removed (🎉) in favor of BBS Git API. @bturner, @gsylvie , I appreciate all the info and help!

One of the last things I experimented with was using `RefService.resolveRef` to identify lightweight and annotated tags. However, I could only get that working for tags that already exist. I had similar experiences with `git for-each-ref`. I ended up using `git cat-file -p` to get tag information for tags being pushed.

**Comment by itay** [comment]
@bturner does the 5.0 SPI expose these variables, or is it still not possible to get access to them such that you can set up a `git` process to take advantage of them (e.g. for `rev-list`)?

**Comment by bturner** [comment]
@itay 

By "these variables" I assume you're referring to the new `GIT_QUARANTINE_PATH` and the `GIT_OBJECT_DIRECTORY` and `GIT_ALTERNATE_OBJECT_DIRECTORIES` that make that all work.

Before I get into accessing the environment, I want to clarify something: If you're forking a `git` process, _you shouldn't need to change your plugin at all_, and you don't need direct access to those environment variables to set your process up, as long as your plugin is using our `CommandBuilder` abstraction. Starting from Bitbucket Server 4.13, whether you're using the "free-form" builder to assemble your own commands from scratch, or you're using the "type-safe" builder (for example, `GitRevListBuilder`), _all_ `Command`s created by the builder will _automatically_ apply the quarantine environment variables necessary when run inside a `pre-receive` hook (the only place the quarantine environment needs to be applied). So the _best_ way forward is to be using our builders to assemble and fork your `git` commands. No plugin should be forking `git` directly.

The only time you might need access to the quarantine environment variables is if you're _not_ forking a `git` process and, instead, you're trying to use something like JGit. To be honest, we're not particularly well-incentivized to do anything to facilitate such an approach. Our support engineers deal with `OutOfMemoryError`s that trace back to JGit's memory usage in third-party add-ons fairly often. The YACC plugin itself has had exactly that sort of issue (see #97, for example). Bitbucket Server does not use JGit _specifically_ because of its memory usage. Forking `git` processes does impose some overhead but a) it keeps all of its memory usage out-of-process to the main Bitbucket Server application and b) while JGit is fast, `git` is generally as fast or faster, depending on the operation.

With all that said, though, the answer to your question is _yes_. In 5.0 there is a new hook SPI which _does_ allow access to _some_ of the `pre-receive` environment variables, and the quarantine variables are among those. The existing `PreReceiveHook` and `PreReceiveRepositoryHook` SPIs are now deprecated and _still do not_ provide access to environment variables. A new `PreRepositoryHook` has been introduced to take their place. (The extremely similar naming won't be ideal during the 5.x release series, but `PreReceiveHook` and `PreReceiveRepositoryHook` will both be removed in 6.0.)

`PreRepositoryHook` is much more than just a replacement for the existing `PreReceiveHook` and `PreReceiveRepositoryHook` SPIs. One of the primary limitations of the existing hooks is that they're 1:1 with Git's `pre-receive` hooks. In some contexts that's desirable, but it means anywhere a `pre-receive` hook wouldn't be invoked, those hooks aren't invoked either. For example, when a pull request is merged, or when branches or tags are created via the UI or REST API, neither `PreReceiveHook`s nor `PreReceiveRepositoryHook`s get invoked. The new `PreRepositoryHook` _will_ be, however. In fact, for _every cancelable ref change_ the base product performs, the new `PreRepositoryHook` is called before it, and the new `PostRepositoryHook` is called after it. Essentially, if a plugin could implement a `PreReceiveHook` or `PreReceiveRepositoryHook`, or listen for some subtype of `RefChangeRequestedEvent`, implementing `PreRepositoryHook` will cover all of those cases (and more) in a single SPI.

Since 5.0 hasn't been released yet the API documentation isn't fully published, but we've uploaded the documentation for a recent release candidate in preparation for when 5.0 is released.
* [`PreRepositoryHook`](https://developer.atlassian.com/static/javadoc/bitbucket-server/5.0.0-rc9/spi/reference/com/atlassian/bitbucket/hook/repository/PreRepositoryHook.html)
* [`RepositoryPushHookRequest`](https://developer.atlassian.com/static/javadoc/bitbucket-server/5.0.0-rc9/api/reference/com/atlassian/bitbucket/hook/repository/RepositoryPushHookRequest.html)
* And, perhaps most relevant to this question, [`ScmHookDetails`](https://developer.atlassian.com/static/javadoc/bitbucket-server/5.0.0-rc9/api/reference/com/atlassian/bitbucket/hook/ScmHookDetails.html)

`ScmHookDetails` is somewhat similar to the existing [`HookResponse`](https://developer.atlassian.com/static/javadoc/bitbucket-server/4.14.4/api/reference/com/atlassian/bitbucket/hook/HookResponse.html), in that it provides access to the `out` and `err` streams, but it also provides access to the environment, and the SCM-specific hook name. The environment is not _complete_ (for several very good reasons, such as the fact that on Windows some environments are MSYS-formatted, like `PATH`, and others aren't), but it has everything a plugin would need to access the quarantine environment.

Again, though. _You really shouldn't need it._ Just use the `CommandBuilder` abstraction to fork your processes.

**Comment by itay** [comment]
@bturner thank you for the extremely detailed response, and apologies to @sford for hijacking this issue for having the discussion!

In our particular case, we use the External Hooks plugin, as we have some custom logic we needed to implement in a pre-receive hook, and we're not a Java shop and were not super comfortable implementing our own plugin we'd have to maintain just to do that logic (for us, we implemented it in Python and are just calling various `git` commands to do the verification we need). 

This functionality obviously ceases to function as of 2.11, so we wanted to know what our recourses were. It sounds like we'll have to wait until 5.0 comes out, and then the plugin gets updated, before we can upgrade the Git version. This is fine, and is certainly on us for using that plugin.

FWIW, the specific check we're doing is a file size limit check during pre-receive, to ensure people can't commit giant files into the repo by accident. The existing plugins that implemented this suffer from huge performance problems which we need to solve given the size of our repository (both in terms of bytes and branches, which are in the thousands). It's been working pretty flawlessly for 2 years now, so we're not in a rush to change it.

Thanks again for the detailed response - I got the info I need, and am now just waiting for 5.0 to be released. 😄
